File: snippets/rust/src/pay_onchain.rs
use std::sync::Arc;

use anyhow::Result;
use breez_sdk_liquid::prelude::*;
use log::info;

async fn get_current_limits(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: get-current-pay-onchain-limits
    let current_limits = sdk.fetch_onchain_limits().await?;

    info!("Minimum amount: {} sats", current_limits.send.min_sat);
    info!("Maximum amount: {} sats", current_limits.send.max_sat);
    // ANCHOR_END: get-current-pay-onchain-limits

    Ok(())
}

async fn prepare_pay_onchain(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-pay-onchain
    let prepare_res = sdk
        .prepare_pay_onchain(&PreparePayOnchainRequest {
            amount: PayOnchainAmount.Receiver {
                amount_sat: 5_000,
            },
            sat_per_vbyte: None,
        })
        .await?;

    // Check if the fees are acceptable before proceeding
    let total_fees_sat = prepare_res.total_fees_sat;
    // ANCHOR_END: prepare-pay-onchain

    Ok(())
}

async fn prepare_pay_onchain_drain(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-pay-onchain-drain
    let prepare_res = sdk
        .prepare_pay_onchain(&PreparePayOnchainRequest {
            amount: PayOnchainAmount.Drain,
            sat_per_vbyte: None,
        })
        .await?;

    // Check if the fees are acceptable before proceeding
    let total_fees_sat = prepare_res.total_fees_sat;
    // ANCHOR_END: prepare-pay-onchain-drain

    Ok(())
}

async fn prepare_pay_onchain_fee_rate(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-pay-onchain-fee-rate
    let optional_sat_per_vbyte = Some(21);

    let prepare_res = sdk
        .prepare_pay_onchain(&PreparePayOnchainRequest {
            amount: PayOnchainAmount.Receiver {
                amount_sat: 5_000,
            },
            sat_per_vbyte: optional_sat_per_vbyte,
        })
        .await?;

    // Check if the fees are acceptable before proceeding
    let claim_fees_sat = prepare_res.claim_fees_sat;
    let total_fees_sat = prepare_res.total_fees_sat;
    // ANCHOR_END: prepare-pay-onchain-fee-rate

    Ok(())
}

async fn start_reverse_swap(
    sdk: Arc<LiquidSdk>,
    prepare_response: PreparePayOnchainResponse,
) -> Result<()> {
    // ANCHOR: start-reverse-swap
    let destination_address = String::from("bc1..");

    sdk.pay_onchain(&PayOnchainRequest {
        address: destination_address,
        prepare_response,
    })
    .await?;
    // ANCHOR_END: start-reverse-swap

    Ok(())
}

File: snippets/go/pay_onchain.go
package example

import (
	"log"

	"github.com/breez/breez-sdk-liquid-go/breez_sdk_liquid"
)

func GetCurrentRevSwapLimits(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: get-current-pay-onchain-limits
	if currentLimits, err := sdk.FetchOnchainLimits(); err == nil {
		log.Printf("Minimum amount, in sats: %v", currentLimits.Send.MinSat)
		log.Printf("Maximum amount, in sats: %v", currentLimits.Send.MaxSat)
	}
	// ANCHOR_END: get-current-pay-onchain-limits
}

func PreparePayOnchain(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-pay-onchain
	amount := breez_sdk_liquid.PayOnchainAmountReceiver{AmountSat: 5_000}
	prepareRequest := breez_sdk_liquid.PreparePayOnchainRequest{
		Amount: amount,
	}

	if prepareResponse, err := sdk.PreparePayOnchain(prepareRequest); err == nil {
		// Check if the fees are acceptable before proceeding
		totalFeesSat := prepareResponse.TotalFeesSat
		log.Printf("Fees: %v sats", totalFeesSat)
	}
	// ANCHOR_END: prepare-pay-onchain
}

func PreparePayOnchainDrain(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-pay-onchain-drain
	amount := breez_sdk_liquid.PayOnchainAmountDrain{}
	prepareRequest := breez_sdk_liquid.PreparePayOnchainRequest{
		Amount: amount,
	}

	if prepareResponse, err := sdk.PreparePayOnchain(prepareRequest); err == nil {
		// Check if the fees are acceptable before proceeding
		totalFeesSat := prepareResponse.TotalFeesSat
		log.Printf("Fees: %v sats", totalFeesSat)
	}
	// ANCHOR_END: prepare-pay-onchain-drain
}

func PreparePayOnchainFeeRate(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-pay-onchain-fee-rate
	amount := breez_sdk_liquid.PayOnchainAmountReceiver{AmountSat: 5_000}
	optionalSatPerVbyte := uint32(21)
	prepareRequest := breez_sdk_liquid.PreparePayOnchainRequest{
		Amount:             amount,
		FeeRateSatPerVbyte: &optionalSatPerVbyte,
	}

	if prepareResponse, err := sdk.PreparePayOnchain(prepareRequest); err == nil {
		// Check if the fees are acceptable before proceeding
		claimFeesSat := prepareResponse.ClaimFeesSat
		totalFeesSat := prepareResponse.TotalFeesSat
		log.Printf("Claim fees: %v sats, total fees: %v sats", claimFeesSat, totalFeesSat)
	}
	// ANCHOR_END: prepare-pay-onchain-fee-rate
}

func StartReverseSwap(sdk *breez_sdk_liquid.BindingLiquidSdk, prepareResponse breez_sdk_liquid.PreparePayOnchainResponse) {
	// ANCHOR: start-reverse-swap
	address := "bc1.."
	payOnchainRequest := breez_sdk_liquid.PayOnchainRequest{
		Address:         address,
		PrepareResponse: prepareResponse,
	}

	if reverseSwapInfo, err := sdk.PayOnchain(payOnchainRequest); err == nil {
		log.Printf("%#v", reverseSwapInfo)
	}
	// ANCHOR_END: start-reverse-swap
}
